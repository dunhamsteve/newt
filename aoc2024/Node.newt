../aoc2023/Node.newt