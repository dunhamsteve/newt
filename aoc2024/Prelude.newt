../aoc2023/Prelude.newt