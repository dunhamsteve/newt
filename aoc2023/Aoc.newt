module Aoc

import Prelude

nums : String → List Int
nums s = map stringToInt $ filter (_/=_ "") $ split (trim s) " "

isDigit : Char -> Bool
isDigit '0' = True
isDigit '1' = True
isDigit '2' = True
isDigit '3' = True
isDigit '4' = True
isDigit '5' = True
isDigit '6' = True
isDigit '7' = True
isDigit '8' = True
isDigit '9' = True
isDigit _ = False

indexOf? : ∀ a. {{Eq a}} → a → List a → Maybe Nat
indexOf? {a} z xs = go Z z xs
  where
    go : Nat → a → List a → Maybe Nat
    go ix z Nil = Nothing
    go ix z (x :: xs) =
      if z == x then Just ix else go (S ix) z xs

-- if_then_else shorthand
-- Lean version uses a decidable instead of Bool
ite : ∀ a. Bool → a → a → a
ite c t e = if c then t else e

-- probably not super efficient, but it works
qsort : ∀ a. (a → a → Bool) → List a → List a
qsort lt Nil = Nil
qsort lt (x :: xs) = qsort lt (filter (λ y => not $ lt x y) xs) ++ x :: qsort lt (filter (lt x) xs)
