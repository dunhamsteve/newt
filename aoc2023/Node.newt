module Node

import Prelude

pfunc fs : JSObject := `require('fs')`
pfunc getArgs : List String := `arrayToList(String, process.argv)`
pfunc readFile uses (fs MkIORes Unit) : (fn : String) -> IO String := `(fn) => (w) => MkIORes(Unit, fs.readFileSync(fn, 'utf8'), w)`
