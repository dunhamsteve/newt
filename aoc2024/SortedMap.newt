module SortedMap

import Prelude

-- TODO We'll want to replace Ord/Eq with (a → Ordering) (and rewrite most of our aoc solutions...)
-- data Ordering : U where
--   LT EQ GT : Ordering

data T23 : Nat -> U -> U -> U where
  Leaf : ∀ k v. k -> v -> T23 Z k v
  Node2 : ∀ h k v. T23 h k v -> k -> T23 h k v -> T23 (S h) k v
  Node3 : ∀ h k v. T23 h k v -> k -> T23 h k v -> k -> T23 h k v -> T23 (S h) k v

lookupT23 : ∀ h k v. {{Ord k}} {{Eq k}} -> k -> T23 h k v -> Maybe (k × v)
lookupT23 key (Leaf k v)= if k == key then Just (k,v) else Nothing
lookupT23 key (Node2 t1 k1 t2) =
  if key <= k1 then lookupT23 key t1 else lookupT23 key t2
lookupT23 key (Node3 t1 k1 t2 k2 t3) =
  if key <= k1 then lookupT23 key t1
  else if key <= k2 then lookupT23 key t2
  else lookupT23 key t3

insertT23 : ∀ h k v. {{Ord k}} {{Eq k}} -> k -> v -> T23 h k v -> Either (T23 h k v) (T23 h k v × k × T23 h k v)
insertT23 key value (Leaf k v) =
  if key == k then Left (Leaf key value)
  else if key <= k then Right (Leaf key value, key, Leaf k v)
  else Right (Leaf k v, k, Leaf key value)
insertT23 key value (Node2 t1 k1 t2) =
  if key <= k1 then
    case insertT23 key value t1 of
      Left t1' => Left (Node2 t1' k1 t2)
      Right (a,b,c) => Left (Node3 a b c k1 t2)
  else case insertT23 key value t2 of
      Left t2' => Left (Node2 t1 k1 t2')
      Right (a,b,c) => Left (Node3 t1 k1 a b c)
insertT23 key value (Node3 t1 k1 t2 k2 t3) =
  if key <= k1 then
    case insertT23 key value t1 of
      Left t1' => Left (Node3 t1' k1 t2 k2 t3)
      Right (a,b,c) => Right (Node2 a b c, k1, Node2 t2 k2 t3)
  else if key <= k2 then
    case insertT23 key value t2 of
      Left t2' => Left (Node3 t1 k1 t2' k2 t3)
      Right (a,b,c) => Right (Node2 t1 k1 a, b, Node2 c k2 t3)
  else
    case insertT23 key value t3 of
      Left t3' => Left (Node3 t1 k1 t2 k2 t3')
      Right (a,b,c) => Right (Node2 t1 k1 t2, k2, Node2 a b c)

-- This is cribbed from Idris. Deleting nodes takes a bit of code.
Hole : Nat → U → U → U
Hole Z k v = Unit
Hole (S n) k v = T23 n k v

Node4 : ∀ k v h. T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → T23 (S (S h)) k v
Node4 t1 k1 t2 k2 t3 k3 t4 = Node2 (Node2 t1 k1 t2) k2 (Node2 t3 k3 t4)

Node5 : ∀ k v h. T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → T23 (S (S h)) k v
Node5 a b c d e f g h i = Node2 (Node2 a b c) d (Node3 e f g h i)

Node6 : ∀ k v h. T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → T23 (S (S h)) k v
Node6 a b c d e f g h i j k = Node2 (Node3 a b c d e) f (Node3 g h i j k)

Node7 : ∀ k v h. T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → k → T23 h k v → T23 (S (S h)) k v
Node7 a b c d e f g h i j k l m = Node3 (Node3 a b c d e) f (Node2 g h i) j (Node2 k l m)

merge1 : ∀ k v h. T23 h k v → k → T23 (S h) k v → k → T23 (S h) k v → T23 (S (S h)) k v
merge1 a b (Node2 c d e) f (Node2 g h i) = Node5 a b c d e f g h i
merge1 a b (Node2 c d e) f (Node3 g h i j k) = Node6 a b c d e f g h i j k
merge1 a b (Node3 c d e f g) h (Node2 i j k) = Node6 a b c d e f g h i j k
merge1 a b (Node3 c d e f g) h (Node3 i j k l m) = Node7 a b c d e f g h i j k l m

merge2 : ∀ k v h. T23 (S h) k v → k → T23 h k v → k → T23 (S h) k v → T23 (S (S h)) k v
merge2 (Node2 a b c) d e f (Node2 g h i) = Node5 a b c d e f g h i
merge2 (Node2 a b c) d e f (Node3 g h i j k) = Node6 a b c d e f g h i j k
merge2 (Node3 a b c d e) f g h (Node2 i j k) = Node6 a b c d e f g h i j k
merge2 (Node3 a b c d e) f g h (Node3 i j k l m) = Node7 a b c d e f g h i j k l m

merge3 : ∀ k v h. T23 (S h) k v → k → T23 (S h) k v → k → T23 h k v → T23 (S (S h)) k v
merge3 (Node2 a b c) d (Node2 e f g) h i = Node5 a b c d e f g h i
merge3 (Node2 a b c) d (Node3 e f g h i) j k = Node6 a b c d e f g h i j k
merge3 (Node3 a b c d e) f (Node2 g h i) j k = Node6 a b c d e f g h i j k
merge3 (Node3 a b c d e) f (Node3 g h i j k) l m = Node7 a b c d e f g h i j k l m

-- height is erased in the data everywhere but the top, but needed for case
-- I wonder if we could use a 1 + 1 + 1 type instead of Either Tree Hole and condense this
deleteT23 : ∀ k v. {{Ord k}} {{Eq k}} → (h : Nat) -> k -> T23 h k v -> Either (T23 h k v) (Hole h k v)
deleteT23 Z key (Leaf k v) = if k == key then Right MkUnit else Left (Leaf k v)
deleteT23 (S Z) key (Node2 t1 k1 t2) =
  if key <= k1
  then case deleteT23 Z key t1 of
    Left t1 => Left (Node2 t1 k1 t2)
    Right _ => Right t2
  else case deleteT23 Z key t2 of
    Left t2 => Left (Node2 t1 k1 t2)
    Right MkUnit => Right t1
deleteT23 (S Z) key (Node3 t1 k1 t2 k2 t3) =
  if key <= k1
  then case deleteT23 _ key t1 of
    Left t1 => Left (Node3 t1 k1 t2 k2 t3)
    Right MkUnit => Left (Node2 t2 k2 t3)
  else if key <= k2 then case deleteT23 _ key t2 of
    Left t2 => Left (Node3 t1 k1 t2 k2 t3)
    Right _ => Left (Node2 t1 k1 t3)
  else case deleteT23 _ key t3 of
    Left t3 => Left (Node3 t1 k1 t2 k2 t3)
    Right _ => Left (Node2 t1 k1 t2)
deleteT23 (S (S h)) key (Node2 t1 k1 t2) =
  if key <= k1
  then case deleteT23 (S h) key t1 of
    Left t1 => Left (Node2 t1 k1 t2)
    Right t1 => case t2 of
      Node2 t2' k2' t3' => Right (Node3 t1 k1 t2' k2' t3')
      Node3 t2 k2 t3 k3 t4 => Left $ Node4 t1 k1 t2 k2 t3 k3 t4
  else case deleteT23 _ key t2 of
    Left t2 => Left (Node2 t1 k1 t2)
    Right t2 => case t1 of
      Node2 a b c => Right (Node3 a b c k1 t2)
      Node3 a b c d e => Left (Node4 a b c d e k1 t2)
deleteT23 (S (S h)) key (Node3 t1 k1 t2 k2 t3) =
  if key <= k1
  then case deleteT23 _ key t1 of
    Left t1 => Left (Node3 t1 k1 t2 k2 t3)
    Right t1 => Left (merge1 t1 k1 t2 k2 t3)
  else if key <= k2 then case deleteT23 _ key t2 of
    Left t2 => Left (Node3 t1 k1 t2 k2 t3)
    Right t2 => Left (merge2 t1 k1 t2 k2 t3)
  else case deleteT23 _ key t3 of
    Left t3 => Left (Node3 t1 k1 t2 k2 t3)
    Right t3 => Left (merge3 t1 k1 t2 k2 t3)

treeLeft : ∀ h k v. T23 h k v → (k × v)
treeLeft (Leaf k v) = (k, v)
treeLeft (Node2 t1 _ _) = treeLeft t1
treeLeft (Node3 t1 _ _ _ _) = treeLeft t1

treeRight : ∀ h k v. T23 h k v → (k × v)
treeRight (Leaf k v) = (k, v)
treeRight (Node2 _ _ t2) = treeRight t2
treeRight (Node3 _ _ _ _ t3) = treeRight t3


data SortedMap : U -> U -> U where
  EmptyMap : ∀ k v. SortedMap k v
  -- not erased so we know what happens in delete
  MapOf : ∀ k v. {h : Nat} →  T23 h k v -> SortedMap k v

deleteMap : ∀ k v. {{Ord k}} {{Eq k}} → k → SortedMap k v → SortedMap k v
deleteMap key EmptyMap = EmptyMap
-- REVIEW if I split h separately in a nested case, it doesn't sort out Hole
deleteMap key (MapOf {k} {v} {Z} tree) = case deleteT23 Z key tree of
  Left t => MapOf t
  Right t => EmptyMap
deleteMap key (MapOf {k} {v} {S n} tree) = case deleteT23 (S n) key tree of
  Left t => MapOf t
  Right t => MapOf t

leftMost : ∀ k v. SortedMap k v → Maybe (k × v)
leftMost EmptyMap = Nothing
leftMost (MapOf m) = Just (treeLeft m)

rightMost : ∀ k v. SortedMap k v → Maybe (k × v)
rightMost EmptyMap = Nothing
rightMost (MapOf m) = Just (treeRight m)

-- TODO issue with metas and case if written as `do` block
pop : ∀ k v. {{Eq k}} {{Ord k}} → SortedMap k v → Maybe ((k × v) × SortedMap k v)
pop m = case leftMost m of
    Just (k,v) => Just ((k,v), deleteMap k m)
    Nothing => Nothing

lookupMap : ∀ k v. {{Ord k}} {{Eq k}} -> k -> SortedMap k v -> Maybe (k × v)
lookupMap k EmptyMap = Nothing
lookupMap k (MapOf map) = lookupT23 k map

updateMap : ∀ k v. {{Ord k}} {{Eq k}} -> k -> v -> SortedMap k v -> SortedMap k v
updateMap k v EmptyMap = MapOf $ Leaf k v
updateMap k v (MapOf map) = case insertT23 k v map of
  Left map' => MapOf map'
  Right (a, b, c) => MapOf (Node2 a b c)

toList : ∀ k v. SortedMap k v → List (k × v)
toList {k} {v} (MapOf smap) = reverse $ go smap Nil
  where
    go : ∀ h. T23 h k v → List (k × v) → List (k × v)
    go (Leaf k v) acc = (k, v) :: acc
    go (Node2 t1 k1 t2) acc = go t2 (go t1 acc)
    go (Node3 t1 k1 t2 k2 t3) acc = go t3 $ go t2 $ go t1 acc
toList _ = Nil
