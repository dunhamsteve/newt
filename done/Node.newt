module Node

import Prelude

pfunc fs : JSObject := `require('fs')`
pfunc getArgs : List String := `arrayToList(String, process.argv.slice(1))`
pfunc readFile uses (fs MkIORes Left Right) : (fn : String) -> IO (Either String String) := `(fn) => (w) => {
  let result
  try {
    let content = fs.readFileSync(fn, 'utf8')
    result = Right(undefined, undefined, content)
  } catch (e) {
    let err = ""+e
    result = Left(undefined, undefined, e)
  }
  return MkIORes(undefined, result, w)
}`

-- I wonder if I should automatically `uses` the constructors in the types
pfunc writeFile uses (fs MkIORes MkUnit) : String → String → IO (Either String Unit) := `(fn, content) => (w) => {
  let result
  try {
    fs.writeFileSync(fn, content, 'utf8')
    result = Right(undefined, undefined, MkUnit)
  } catch (e) {
      let err = ""+e
      result = Left(undefined, undefined, e)
  }
  return MkIORes(undefined, result, w)
}`

-- maybe System.exit or something, like the original putStrLn msg >> exitFailure
pfunc exitFailure : ∀ a. String → a := `(_, msg) => { 
  console.log(msg);
  process.exit(1);
}`

