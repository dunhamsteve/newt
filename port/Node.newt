module Node

import Prelude

pfunc fs : JSObject := `require('fs')`
pfunc getArgs : List String := `arrayToList(String, process.argv.slice(1))`
pfunc readFile uses (fs MkIORes Left Right) : (fn : String) -> IO (Either String String) := `(fn) => (w) => {
  let result
  try {
    let content = fs.readFileSync(fn, 'utf8')
    result = Right(null, null, content)
  } catch (e) {
    let err = ""+e
    result = Left(null, null, e)
  }
  return MkIORes(null, result, w)
}`

-- I wonder if I should automatically `uses` the constructors in the types
pfunc writeFile uses (fs MkIORes MkUnit) : String → String → IO (Either String Unit) := `(fn, content) => (w) => {
  let result
  try {
    fs.writeFileSync(fn, content, 'utf8')
    result = Right(null, null, MkUnit)
  } catch (e) {
      let err = ""+e
      result = Left(null, null, e)
  }
  return MkIORes(null, result, w)
}`

-- maybe System.exit or something, like the original putStrLn msg >> exitFailure
pfunc exitFailure : ∀ a. String → a := `(_, msg) => { 
  console.log(msg);
  process.exit(1);
}`

