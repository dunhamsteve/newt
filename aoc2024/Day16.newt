module Day16

import Prelude
import Node
import Aoc
import SortedMap

gridPoints : String → List (Char × Int × Int)
gridPoints text = go 0 0 (unpack text) Nil
  where
    go : Int → Int → List Char → List (Char × Int × Int) → List (Char × Int × Int)
    go row col Nil points = points
    go row col ('\n' :: cs) points = go (row + 1) 0 cs points
    go row col (c :: cs) points = go row (col + 1) cs ((c,row,col) :: points)

Grid : U → U
Grid a = SortedMap Point a

getGrid : ∀ a. String → (Char → a) → Grid a
getGrid {a} text f = foldl update EmptyMap $ gridPoints text
  where
    update : Grid a → Char × Point → Grid a
    update grid (c, pt) = updateMap pt (f c) grid

find : ∀ a. (a → Bool) → List a → Maybe a
find f Nil = Nothing
find f (x :: xs) = if f x then Just x else find f xs

fromList : ∀ k v. {{Ord k}} {{Eq k}} → List (k × v) → SortedMap k v
fromList xs = foldMap (\ a b => b) EmptyMap xs

data Dir : U where
  North South East West : Dir

-- all of our co

dist : Point → Point → Int
dist (a,b) (c,d) =
  let dr = if a < b then b - a else a - b
      dc = if c < d then d - c else c - d
  in dr + dc

dirs : List Dir
dirs = (North :: South :: East :: West :: Nil)

turn : Dir → Dir → Int
turn North North = 0
turn South South = 0
turn East East = 0
turn West West = 0
turn North South = 2000
turn South North = 2000
turn East West = 2000
turn West East = 2000
turn _ _ = 1000

DPoint : U
DPoint = Point × Dir

move : DPoint → DPoint
move ((r,c), North) = ((r - 1, c), North)
move ((r,c), South) = ((r + 1, c), South)
move ((r,c), East) = ((r, c + 1), East)
move ((r,c), West) = ((r, c - 1), West)

unmove : DPoint → DPoint
unmove ((r,c), South) = ((r - 1, c), North)
unmove ((r,c), North) = ((r + 1, c), South)
unmove ((r,c), West) = ((r, c + 1), East)
unmove ((r,c), East) = ((r, c - 1), West)

dirVal : Dir → Int
dirVal North = 0
dirVal South = 1
dirVal East = 2
dirVal West = 3

instance Ord Dir where
  a < b = dirVal a < dirVal b

instance Eq Dir where
  a == b = dirVal a == dirVal b

instance Eq DPoint where
  (a,b) == (c,d) = a == c && b == d

instance Ord DPoint where
  (a,b) < (c,d) = a < c || a == c && b < d


Cand : U
Cand = Int × Point × Dir

instance Eq Cand where
  (s,pt) == (s',pt') = s == s' && pt == pt'

instance Ord Cand where
  (s,pt) < (s',pt') = s < s' || s == s' && pt < pt'

min : Int → Int → Int
min x y = if x < y then x else y

-- todo is (est,dpoint) -> cost
-- scores helps us cut (dpoint -> cost) -- cost or est should be the same
-- We return the score map for part2 (which reconstructs the optimal paths)
part1 : Point → SortedMap DPoint Int → SortedMap Cand Int → SortedMap DPoint Int
part1 end scores todo =
  let (Just (((est, pt),cost), todo)) =  (pop todo) | Nothing => scores in
  let (Just (_,best)) = lookupMap pt scores | _ => part1 end scores todo in
  if best < cost then part1 end scores todo else
  let scores = updateMap pt cost scores in
  -- keep going so we collect all optimal paths
  if fst pt == end then part1 end scores todo else
  let todo = foldMap min todo $ mapMaybe (next pt cost) dirs in
  part1 end scores todo
  where
    next : Point × Dir → Int → Dir → Maybe (Cand × Int)
    next dp@(pt,dir) cost newdir = do
      let cost = cost + turn dir newdir + 1
      let dp' = move (pt, newdir)
      case lookupMap dp' scores of
        Nothing => Nothing
        Just (_, best) => if best < cost
          then Nothing
          else Just ((cost + dist (fst dp') end, dp'), cost)

-- work backwards to collect optimal path
goBack : SortedMap DPoint Int → List (DPoint × Int) → SortedMap Point Unit → List Point
goBack scores Nil tiles = map fst $ toList tiles
goBack scores ((dp@(pt,dir),cost) :: todo) tiles =
  let tiles = updateMap (fst dp) MkUnit tiles in
  let next = filter valid $ mapMaybe (flip lookupMap scores ∘ (_,_ (fst $ unmove dp))) dirs in
  goBack scores (next ++ todo) tiles
  where
    -- if the costs add up, this link is part of the path
    valid : DPoint × Int → Bool
    valid  cand@((pt', dir'), cost') = cost == cost' + 1 + turn dir dir'

run : String -> IO Unit
run fn = do
  putStrLn fn
  text <- readFile fn
  let pts = filter (_/=_ '#' ∘ fst) $ gridPoints text
  let (Just (_, start)) = find (_==_ 'S' ∘ fst) pts | _ => putStrLn "no start"
  let (Just (_,end)) = find (_==_ 'E' ∘ fst) pts | _ => putStrLn "no end"

  let scores = fromList $ join $ map (\ x => map ( \d => ((snd x, d), 999999)) dirs) pts
  let todo = updateMap (0, start, East) 0 EmptyMap
  let scores = part1 end scores todo
  let (ends@((_, e) :: es)) = mapMaybe (flip lookupMap scores) $ map (_,_ end) dirs | _ => putStrLn "no end"
  let p1 = foldl min e $ map snd es
  putStrLn $ "part1 " ++ show p1
  let todo = filter (_==_ p1 ∘ snd) ends
  let tiles = goBack scores todo EmptyMap
  putStrLn $ "part2 " ++ show (length tiles)

main : IO Unit
main = do
  run "aoc2024/day16/eg.txt"
  run "aoc2024/day16/eg2.txt"
  run "aoc2024/day16/input.txt"
