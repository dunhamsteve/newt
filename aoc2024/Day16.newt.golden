aoc2024/day16/eg.txt
part1 7036
part2 45
aoc2024/day16/eg2.txt
part1 11048
part2 64
aoc2024/day16/input.txt
part1 90460
part2 575
